packages/delivery-expo/queue.js

```diff
@@ -0,0 +1,143 @@
+const { FileSystem } = require('expo')
+
+const MAX_ITEMS = 64
+const PAYLOAD_PATH = `${FileSystem.cacheDirectory}bugsnag`
+const filenameRe = /^bugsnag-.*\.json$/
+
+/*
+ * This class resembles FIFO queue in which to store undelivered payloads.
+ */
+module.exports = class UndeliveredPayloadQueue {
+  constructor (resource, onerror = () => {}) {
+    this._resource = resource
+    this._path = `${PAYLOAD_PATH}/${this._resource}`
+    this._onerror = onerror
+    this._truncating = false
+  }
+
+  /*
+   * Ensure the persistent cache directory exists
+   */
+  async init () {
+    const { exists, isDirectory } = await FileSystem.getInfoAsync(this._path)
+    if (exists && isDirectory) return true
+    await FileSystem.makeDirectoryAsync(this._path, { intermediates: true })
+  }
+
+  /*
+   * Keeps the queue size bounded by MAX_LENGTH
+   */
+  async _truncate () {
+    // this isn't atomic so only enter this method once at any time
+    if (this._truncating) return
+    this._truncating = true
```

Comment:
JS is single threaded 😁

Reply:
Maybe we can talk this through tomorrow - I'm not entirely sure why there's a need for a flag if concurrent access to the method isn't expected. Putting it in the terms of threads was probably a bit misleading on my part

Reply:
Concurrent access is expected, but js only switches execution contexts on i/o and so it will only execute another context later in this function when the promises are used to read the filesystem. So you can maybe think of this part as atomically executed

Reply:
That makes sense, I don't think there's anything to address here in that case.